src/parser/types.rs

```diff
@@ -38,15 +42,65 @@ pub fn set(input: &str) -> IResult<&str, Vec<String>, VerboseError<&str>> {
     })
 }
 
-fn string(i: &str) -> IResult<&str, &str, VerboseError<&str>> {
+pub fn hashmap(input: &str) -> IResult<&str, HashMap<String, Types>, VerboseError<&str>> {
+    context(
+        "map",
+        preceded(
+            char('{'),
+            cut(terminated(
+                map(
+                    separated_list0(preceded(sp, char(',')), key_value),
+                    |tuple_vec| {
+                        tuple_vec
+                            .into_iter()
+                            .map(|(k, v)| (String::from(k), v))
+                            .collect()
+                    },
+                ),
+                preceded(sp, char('}')),
+            )),
+        ),
+    )(input)
+}
+
+fn key_value(input: &str) -> IResult<&str, (&str, Types), VerboseError<&str>> {
+    separated_pair(
+        preceded(sp, alphanumerickey),
+        cut(preceded(sp, char(':'))),
+        wql_value,
+    )(input)
+}
+
+pub fn vector(input: &str) -> IResult<&str, Vec<Types>, VerboseError<&str>> {
+    context(
+        "vector",
+        preceded(
+            tag("["),
+            cut(terminated(
+                separated_list0(preceded(char(','), sp), wql_value),
+                alt((char(']'), preceded(sp, char(']')))),
```

Comment:
Why is this needed (two `]`)?

Reply:
`alt` gives an alternative parser choice, `char(])` looks only for a terminating element of char `]`, while `preceded` with `sp`(spaces) looks for a any terminating element consisting of whitespaces and a char `]`.

Ex:
- only `char(']')` would parse this `[1, 2, 3, 4]` but not `[1, 2, 3, 4 ]` or `[1, 2, 3, 4.         ]`
- only `preceded(sp, char(']'))` would parse `[1, 2, 3, 4 ]` or `[1, 2, 3, 4.         ]` but sometimes it bugged with `[1, 2, 3, 4]`